File: bridges/android/location-tracker/src/main/java/co/zw/nissangtr/bridges/location/GpsModels.kt
package co.zw.nissangtr.bridges.location

/**
 * Mirrors bridges/contracts/gps.ts — native Kotlin surface for GpsBridge.
 * Bridge emits coordinates only; callers map via [toDeliveryLocationIngest]
 * and post with RPC ingest_delivery_location (never from this module).
 */

enum class LocationPermissionStatus {
    GRANTED,
    DENIED,
    RESTRICTED,
    NOT_DETERMINED,
    /** Approximate-only when fine was not granted. */
    APPROXIMATE,
}

/** One native position fix from the device. */
data class GpsCoordinate(
    val latitude: Double,
    val longitude: Double,
    /** Horizontal accuracy in meters when the OS provides it. */
    val accuracyMeters: Float? = null,
    /** ISO-8601 timestamp from the device (maps to p_recorded_at). */
    val capturedAt: String,
)

/**
 * Payload shape for ingest_delivery_location after a bridge fix.
 * Field names mirror RPC parameters for management-app / service callers.
 */
data class DeliveryLocationIngest(
    val deliveryJobId: String,
    val lat: Double,
    val lng: Double,
    /** ISO-8601; omit to let the RPC default to now(). */
    val recordedAt: String? = null,
    /** Meters; omit when accuracy unknown. */
    val accuracyM: Double? = null,
)

fun toDeliveryLocationIngest(
    deliveryJobId: String,
    coord: GpsCoordinate,
): DeliveryLocationIngest = DeliveryLocationIngest(
    deliveryJobId = deliveryJobId,
    lat = coord.latitude,
    lng = coord.longitude,
    recordedAt = coord.capturedAt,
    accuracyM = coord.accuracyMeters?.toDouble(),
)

interface GpsWatchHandle {
    suspend fun stop()
}

/**
 * Native GPS for delivery tracking.
 * Prefer [watchPosition] while a job is en route; throttle client-side (~5s)
 * before calling ingest_delivery_location. This bridge does not call Supabase.
 */
interface GpsBridge {
    suspend fun getLocationPermissionStatus(): LocationPermissionStatus
    suspend fun requestLocationPermission(): LocationPermissionStatus
    /** One-shot current position (permission-gated on device). */
    suspend fun getCurrentPosition(): GpsCoordinate
    /**
     * Stream updates for delivery tracking; caller must [GpsWatchHandle.stop].
     * Starts a location foreground service so tracking can continue when backgrounded.
     */
    suspend fun watchPosition(
        onUpdate: (GpsCoordinate) -> Unit,
        onError: ((String) -> Unit)? = null,
    ): GpsWatchHandle
}

File: bridges/ios/LocationTracker/Sources/LocationTracker/GpsModels.swift
import Foundation

/// Mirrors bridges/contracts/gps.ts — native Swift surface for GpsBridge.
/// Bridge emits coordinates only; callers map via `toDeliveryLocationIngest`
/// and post with RPC `ingest_delivery_location` (never from this module).

public enum LocationPermissionStatus: String, Sendable {
    case granted
    case denied
    case restricted
    case notDetermined = "not_determined"
    /// Android approximate-only; on iOS unused (maps to denied/notDetermined).
    case approximate
}

/// One native position fix from the device.
public struct GpsCoordinate: Sendable, Equatable {
    public var latitude: Double
    public var longitude: Double
    /// Horizontal accuracy in meters when the OS provides it.
    public var accuracyMeters: Double?
    /// ISO-8601 timestamp from the device (maps to p_recorded_at).
    public var capturedAt: String

    public init(
        latitude: Double,
        longitude: Double,
        accuracyMeters: Double? = nil,
        capturedAt: String
    ) {
        self.latitude = latitude
        self.longitude = longitude
        self.accuracyMeters = accuracyMeters
        self.capturedAt = capturedAt
    }
}

/// Payload shape for `ingest_delivery_location` after a bridge fix.
public struct DeliveryLocationIngest: Sendable, Equatable {
    public var deliveryJobId: String
    public var lat: Double
    public var lng: Double
    public var recordedAt: String?
    public var accuracyM: Double?

    public init(
        deliveryJobId: String,
        lat: Double,
        lng: Double,
        recordedAt: String? = nil,
        accuracyM: Double? = nil
    ) {
        self.deliveryJobId = deliveryJobId
        self.lat = lat
        self.lng = lng
        self.recordedAt = recordedAt
        self.accuracyM = accuracyM
    }
}

public func toDeliveryLocationIngest(
    deliveryJobId: String,
    coord: GpsCoordinate
) -> DeliveryLocationIngest {
    DeliveryLocationIngest(
        deliveryJobId: deliveryJobId,
        lat: coord.latitude,
        lng: coord.longitude,
        recordedAt: coord.capturedAt,
        accuracyM: coord.accuracyMeters
    )
}

public protocol GpsWatchHandle: AnyObject {
    func stop() async
}

/// Native GPS for delivery tracking.
/// Prefer `watchPosition` while a job is en route; throttle client-side (~5s)
/// before calling `ingest_delivery_location`. This bridge does not call Supabase.
public protocol GpsBridge: AnyObject {
    func getLocationPermissionStatus() async -> LocationPermissionStatus
    func requestLocationPermission() async -> LocationPermissionStatus
    func getCurrentPosition() async throws -> GpsCoordinate
    func watchPosition(
        onUpdate: @escaping @Sendable (GpsCoordinate) -> Void,
        onError: (@Sendable (String) -> Void)?
    ) async throws -> GpsWatchHandle
}
